APS-1536 Add new seed from Excel job.

danhumphreys-moj opened the pull request:
We want the ability to be able to update approved premises room characteristic data from an Excel spreadsheet.  This commit adds the functionality to run seed jobs from xlsx files and populated rooms, beds and characteristics.

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+          characteristics.add(characteristic)
+          log.info("Added characteristic ${characteristic.propertyName} to room code $roomCode.")
+        } else {
+          if (characteristics.remove(characteristic)) {
```
Thinking to delete all characteristics and then re-create from the Excel file.  Would need integration to check that we rollback if there are any problems.

davidatkinsuk replied:
Yep makes sense

danhumphreys-moj replied:
Logic updated to delete and re-create characteristics.  Ticket [APS-1732](https://dsdmoj.atlassian.net/browse/APS-1732) raised to add integration test to check that transactions are rolled back if seed job fails at the database stage.

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+
+      val characteristics = mutableListOf<CharacteristicEntity>()
+
+      questions.drop(2).forEachIndexed { i, question ->
```
Spent some time looking at whether we can go straight from dataframe to list of objects but not easy with the layout of the site survey data as it stands.  It would require characteristics as column headers and room ids as row indices.  So a transposition of the existing layout (room ids as column headers and characteristics as row indices).

More straightforward to loop through the columns (for now).

danhumphreys-moj replied:
No need to investigate this - seed job works as is.

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+
+      questions.drop(2).forEachIndexed { i, question ->
+        val characteristic = siteSurvey.questionToCharacterEntityMapping.getValue(question.toString())!!
+        if (roomCharacteristics[i + 2].toString().equals("yes", true)) {
```
First two rows in the Excel spreadsheet are room and bed ids so need to be dropped when comparing to characteristic data. Will add a comment in the code.

davidatkinsuk replied:
I think it may be cleaner to deal with the spreadsheet in two stages

Stage 1 - load the data into an internal model (use data classes). Do validation whilst loading. The model should use booleans instead of "Yes/No" strings
Stage 2 - use the internal model to update premise/room/bed

That way all of the awkward use of indexes is used in one place and should need less documentation?

danhumphreys-moj replied:
Logic has been restructured to do this.  Load, validate and create rooms, beds and characteristics, then save to database.

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+@Component
+class SiteSurvey constructor(characteristicRepository: CharacteristicRepository) {
+  private val questionToPropertyNameMapping = mapOf(
+//    "Is this bed in a single room?" to "isSingle",
```
Will uncomment when we get to running for all characteristics.

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
Originally this class was an object with a static property but seemed more appropriate to have this mapping and since this required using characteristicRepository used @Component and getBean() which seems to work.

danhumphreys-moj commented on src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/cas1/seed/SeedFromXLSXApprovedPremisesRoomsTest.kt:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  @Test
+  fun `Removing a characteristic that exists succeeds`() {
```
Redundant if we remove all characteristics to begin with.

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+      val bedCode = roomCharacteristics[1].toString()
+      val bedName = "$roomName - $bedCode"
+
+      val bed = room?.beds?.find { it.code == bedCode } ?: createBed(
```
Beds aren't actually used during this job, however I have presumed that if they are in the spreadsheet and not in the database then we need to create them as part of this import task.

davidatkinsuk replied:
agreed

danhumphreys-moj replied:
Beds are created if they don't exist.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/SeedJob.kt:
```diff
@@ -27,4 +28,18 @@ abstract class SeedJob<RowType>(
+    }
+  }
+
+  abstract fun processDataFrame(dataFrame: DataFrame<*>)
```
Again, one for once everything is working but i think seed jobs should be spring managed beans and whatever arguments are required should be passed into the function doing the processing. I'm slowly working on changing the existing seed jobs to use this pattern

danhumphreys-moj replied:
This has been changed for all seed jobs now.  Have followed the same principle for this job.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+    findExistingPremisesOrThrow(premisesId)
+
+    val questions = dataFrame.getColumn(0).values().toList()
+    if (questions.drop(2) != siteSurvey.questionToCharacterEntityMapping.keys.toList()) throw RuntimeException("Site survey file $fileName contains wrong questions on sheet $sheetName")
```
Do we actually care if the number of questions mismatch, as long as we can find the rows we're interested in? It's easy in excel to add additional 'blank' rows and i think this check will fail in that case?

danhumphreys-moj replied:
Have removed this check.  Updated the logic so that we run through the list of questions we expect to see in the site survey data.  If we do not find a match in the premise spreadsheet then the import fails.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+      )
+
+      val bedCode = roomCharacteristics[1].toString()
+      val bedName = "$roomName - $bedCode"
```
I think we should be using the value of the 'Room Number / Name' row for this. I _think_ this is the first time premises have been able to define their own names

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+}
+
+@Component
+class SiteSurvey constructor(characteristicRepository: CharacteristicRepository) {
```
i don't _think_ you need the constructor keyword here?

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
I think this should be probably be a function so we resolve the characteristics everytime they are required (e.g. get hold of this list once in the seed job when processing the excel). These things rarely change but they are changed directly in the DB this list will be stale

danhumphreys-moj replied:
Raised ticket [APS-1734](https://dsdmoj.atlassian.net/browse/APS-1734) to do this.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,166 @@
+
+      questions.drop(2).forEachIndexed { i, question ->
+        val characteristic = siteSurvey.questionToCharacterEntityMapping.getValue(question.toString())!!
+        if (roomCharacteristics[i + 2].toString().equals("yes", true)) {
```
I think we need should trim and lower case any cell we read before comparisons as humans are likely to add junk in the fields. I also wonder if we should error if something other than "yes" or "no" is in the cell e.g. if someone added "sometimes" :)

danhumphreys-moj replied:
Added `.trim()`; `true` also passed to equals which ignores case.

Discussed addition of "N/A" to string comparison logic with @davidatkinsuk this morning.  If answer is "N/A" - answer is set to "No".

danhumphreys-moj commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/SeedFromExcelController.kt:
```diff
@@ -0,0 +1,19 @@
+  override fun seedFromExcelPost(seedFromExcelRequest: SeedFromExcelRequest): ResponseEntity<Unit> {
+    throwIfNotLoopbackRequest()
+
+    seedService.seedExcelData(seedFromExcelRequest.seedType, seedFromExcelRequest.premisesId, seedFromExcelRequest.fileName)
```
I was thinking to pass the premise id as an argument to the run seed script to begin with (haven't done this yet), seemed the most straightforward thing to do.

Then run the script for every spreadsheet rather than for every file in a directory?

davidatkinsuk replied:
Probably one to return to later. We have 108 premises, most of which (presumably) we'd want to load in bulk. We could always just script a curl command to be fair!

danhumphreys-moj replied:
Ticket [APS-1733](https://dsdmoj.atlassian.net/browse/APS-1733) raised to investigate simplifying the existing script.

davidatkinsuk commented on src/main/resources/static/_shared.yml:
```diff
@@ -3719,6 +3733,10 @@ components:
+    SeedFromExcelFileType:
+      type: string
+      enum:
+        - approved_premises_room_from_excel
```
can drop _from_excel here i think? given it's of type `SeedFromExcelFileType`

danhumphreys-moj replied:
Have removed `_from_excel`.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
I _think_ these are all premise characteristics? In which case they're safe to ignore and can probably remove this comment

danhumphreys-moj replied:
Removed these commented characteristics now.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    // update the rooms
+    rooms = createOrUpdateRooms(rooms)
```
I think the code is nicely organised so we don't need these comments anymore

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+  private val siteSurvey: SiteSurvey,
+) : ExcelSeedJob() {
+  private val log = LoggerFactory.getLogger(this::class.java)
+  override fun processDataFrame(dataFrame: DataFrame<*>, premisesId: UUID) {
```
would a better name for the dataFrame variable be 'roomsWorksheet?'

danhumphreys-moj replied:
Renamed to `roomsWorksheet`.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+
+    for (i in 1..<dataFrame.columnsCount()) {
+      val roomAnswers = dataFrame.getColumn(i)
+      val room = createRoom(premisesId, roomAnswers.name, roomAnswers[0].toString())
```
i think what these variables are (especially `roomAnswers[0].toString()`) would be clearer if you provided the function parameter names?

danhumphreys-moj replied:
Had added parameter names for clarity.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+    characteristics = mutableListOf(),
+  )
+
+  fun createBed(bedName: String, bedCode: String, room: RoomEntity): BedEntity = BedEntity(
```
i think for consistency with code earlier on this should be called `buildBed` (as we're not actually creating in the database at this point?)

danhumphreys-moj replied:
Renamed to `buildBed`.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+    }
+  }
+
+  private fun createRoom(premisesId: UUID, roomCode: String, roomName: String): RoomEntity = RoomEntity(
```
i think for consistency with code earlier on this should be called `buildRoom` (as we're not actually creating in the database at this point?)

danhumphreys-moj replied:
Renamed to `buildRoom`.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+    siteSurvey.questionToCharacterEntityMapping.forEach { (question, characteristic) ->
+      val rowId = dataFrame.getColumn(0).values().indexOf(question)
+
+      if (rowId == -1) throw SiteSurveyImportException("Characteristic question '$question' not found on sheet Sheet3.")
```
nice

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+        val roomCode = dataFrame.getColumn(colId).name
+        val answer = dataFrame[rowId][colId].toString().trim()
+
+        if (answer.equals("yes", true)) {
```
could you please adding a parameter name for the second argument (ignore) for readibility

danhumphreys-moj replied:
Added parameter name.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
I think this needs to error if a characteristic can't be found e.g.

`characteristicRepository.findByPropertyName(value, ServiceName.approvedPremises.value)!!`

This will then remove the need to use `!!` when retrieving characteristics from here

davidatkinsuk replied:
It's probably neater to do something like

```
characteristicRepository.findByPropertyName(value, ServiceName.approvedPremises.value) ?: error("Can't find characteristic")
```

Note that because this is being ran on spring startup this will stop the service from starting. This will be resolved by APS-1734

davidatkinsuk replied:
I was thinking about this last night and I think it's probably best to leave this until working on APS-1734 to avoid the possibility of breaking the entire service if a characteristic is missing :)

danhumphreys-moj replied:
Ok, no problem.  Will mark as resolved since will be addressed by APS-1734.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+        if (answer.equals("yes", true)) {
+          premisesCharacteristics.computeIfAbsent(roomCode) { mutableListOf() }.add(characteristic!!)
+        } else if (!answer.equals("no", true) && !answer.equals("N/A", true)) {
+          throw SiteSurveyImportException("Expecting 'yes' or 'no' for question '$question' but is '$answer' on sheet Sheet3.")
```
great! it would be helpful to output rowId and colId here too?

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+          "${roomAnswers.name} - $bedNumber",
+          rooms.firstOrNull { it.code == roomAnswers.name }
+            ?: throw IllegalArgumentException("Room not found with id ${roomAnswers.name} and code ${roomAnswers.name}."),
+        ),
```
can you please add function param names here so it's clear what each value represents

danhumphreys-moj replied:
Added parameter names.

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+        persistedRoom.characteristics.clear()
+        persistedRoom.characteristics.addAll(room.characteristics)
+        roomRepository.save(persistedRoom)
+        log.info("Added characteristic(s) ${persistedRoom.characteristics.joinToString()} to room code ${persistedRoom.code}.")
```
nice!

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+    beds.forEach {
+      val existingBed = bedRepository.findByCodeAndRoomId(it.code!!, it.room.id)
+      if (existingBed != null) {
+        log.info("Bed ${it.id} with code ${it.code} already exists in room code ${it.room.code}.")
```
Something to to via another ticket - we need to decide whether to update any fields on existing rooms.

danhumphreys-moj replied:
If a room with the same code exists, the characteristics are cleared out and the ones from the site survey spreadsheet are used.
If a room doesn't exist with this code, it is created with the characteristics from the site survey data.

If a bed exists with the same code then nothing is done (apart from logging that this has happened).
If a bed doesn't exist with the same code then one is created.

danhumphreys-moj replied:
Ahh ok - I see.  I wondered if this is what you meant.  Have raised [APS-1735](https://dsdmoj.atlassian.net/browse/APS-1735) to determine if this is a requirement.

davidatkinsuk commented on src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/cas1/seed/SeedFromXLSXApprovedPremisesRoomsTest.kt:
```diff
@@ -0,0 +1,436 @@
+    assertThat(newBed!!.name).isEqualTo("SWABI01NEW")
+    assertThat(
+      newBed.room.id == newRoom.id &&
+        newBed.room.code == newRoom.code,
```
I think when we assert on `newBed.room.code` we can use a String literal? Same goes for anywhere else doing a similar assertions

danhumphreys-moj replied:
Changed to use string literals where applicable.

davidatkinsuk reviewed (approved):
LGTM and thanks for the thorough testing 👍

davidatkinsuk commented on src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt:
```diff
@@ -0,0 +1,210 @@
+  )
+
+  private fun findExistingPremisesOrThrow(premisesId: UUID): PremisesEntity {
+    return premisesRepository.findByIdOrNull(premisesId) ?: throw SiteSurveyImportException(
```
could you please change this use approvedPremisesRepository instead? That way we won't accidentally update CAS3 premises

danhumphreys-moj replied:
Done!

danhumphreys-moj commented:
Addressed all PR comments where I could, and raised tickets for others that can be done after we run this seed job in pre-production.  Tested updated branch locally again and site survey import was successful.

davidatkinsuk commented:
> Addressed all PR comments where I could, and raised tickets for others that can be done after we run this seed job in pre-production. Tested updated branch locally again and site survey import was successful.

Great, merge at will 👍
